File: ponio/examples/lotka_volterra.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// NOLINTBEGIN(misc-include-cleaner)

#include <filesystem>
#include <string>
#include <valarray>

#include <ponio/observer.hpp>
#include <ponio/problem.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>
#include <ponio/time_span.hpp>

#include <CLI/CLI.hpp>

// NOLINTEND(misc-include-cleaner)

/*
Lotka-Volterra system
---------------------

$$
  \begin{cases}
    \frac{\mathrm{d}x}{\mathrm{d}t} = \alpha x - \beta xy\\
    \frac{\mathrm{d}y}{\mathrm{d}t} = \delta xy - \gamma y\\
  \end{cases}
$$

with $\alpha = \frac{2}{3}$, $\beta = \frac{4}{3}$, and $\delta = \gamma = 1$. Initial condition $(x,y)(t=0) = (x_0,x_0)$ done by user.

This system is solved by RK(11,8) Runge-Kutta method with time step $\Delta t=0.1$ to the time $t\leq 15$.
 */

int
main( int argc, char** argv )
{
    CLI::App app{ "Launch Lotka-Volerra equation simulation solved by RK(11,8) method" }; // NOLINT(misc-include-cleaner)

    // default filename
    std::string const dirname = "lv_data";
    auto filename             = std::filesystem::path( dirname ) / "lv.dat";

    using state_t = std::valarray<double>;

    double x0 = 1.0; // default x0

    app.add_option( "filename", filename, "name of output file" );
    app.add_option( "x0", x0, "initial condition (x,y)(t=0) = (x0,x0)" );
    CLI11_PARSE( app, argc, argv ); // NOLINT(misc-include-cleaner)

    ponio::observer::file_observer fobs( filename );

    // parameters
    double const alpha = 2. / 3.;
    double const beta  = 4. / 3.;
    double const gamma = 1.;
    double const delta = 1.;

    auto lotka_volterra_pb = ponio::make_simple_problem( // define problem
        [=]( double, state_t const& u ) -> state_t
        {
            return { alpha * u[0] - beta * u[0] * u[1], delta * u[0] * u[1] - gamma * u[1] };
        } );

    ponio::time_span<double> const t_span = { 0., 15. }; // begin and end time

    double const dt  = 0.1;        // time step
    state_t const u0 = { x0, x0 }; // initial condition

    ponio::solve( lotka_volterra_pb, ponio::runge_kutta::rk_118(), u0, t_span, dt, fobs );

    return 0;
}

File: ponio/examples/lotka_volterra.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import dataclasses
import argparse

import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt

name = "lotka_volterra"
data_dir = name+"_data"
filename_fmt = f"{name}_{{:0.2f}}.dat"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)


@dataclasses.dataclass
class simu_data:
    x0: float
    data: npt.NDArray[np.float64]

    def time(self):
        return self.data[:, 0]

    def rabbit(self):
        return self.data[:, 1]

    def fox(self):
        return self.data[:, 2]


def load_data(x0):
    return simu_data(x0=x0, data=np.loadtxt(os.path.join(data_dir, filename_fmt.format(x0))))


os.makedirs(data_dir, exist_ok=True)

make = subprocess.Popen(["make", name])
make.wait()

X0 = np.linspace(0.9, 1.8, 10)

# launch multiple simulation
for x0 in X0:
    args = [os.path.join(".", name),
            os.path.join(data_dir, filename_fmt.format(x0)), str(x0)]
    process = subprocess.Popen(args)
    process.wait()

data = [load_data(x0) for x0 in X0]

# composantes plot
fig, (ax1, ax2) = plt.subplots(nrows=2, figsize=(10, 6))

for i, d in enumerate(data):
    ax1.plot(d.time(), d.rabbit(), color=f"C{i}", label=f"$x_0 = {d.x0:0.2f}$")
    ax2.plot(d.time(), d.fox(), color=f"C{i}", label=f"$x_0 = {d.x0:0.2f}$")

ax1.set_ylabel("prey: $x$")
ax2.set_ylabel("predator: $y$")
ax2.set_xlabel("time")

box1 = ax1.get_position()
ax1.set_position([box1.x0, box1.y0, box1.width * 0.95, box1.height])
box2 = ax2.get_position()
ax2.set_position([box2.x0, box2.y0, box2.width * 0.95, box2.height])

ax1.legend(bbox_to_anchor=(1.01, 1),
           loc='upper left', borderaxespad=0.)

plt.savefig(os.path.join(img_dir, "01.png"))

if not arguments.only_save:
    fig.suptitle("Lotka-Volterra system (solved witk RK (11,8))")
    plt.show()


for i, d in enumerate(data):
    plt.plot(d.rabbit(), d.fox(), color=f"C{i}", label=f"$x_0 = {d.x0:0.2f}$")
    plt.plot([d.x0], [d.x0], "o", color=f"C{i}", label=None)

plt.xlabel("$x$")
plt.ylabel("$y$")
plt.legend()

plt.savefig(os.path.join(img_dir, "02.png"))

if not arguments.only_save:
    plt.title("Lotka-Volterra system (solved witk RK (11,8))")
    plt.show()
